src/app/convai-checker.component.spec.ts

```diff
@@ -1455,4 +1455,451 @@ describe('Convai checker test', () => {
     // Send an input event to trigger the service call.
     setTextAndFireInputEvent(queryTexts[callCount], textArea);
   }));
+
+  it('Test circle square diamond change for score thresholds', async(() => {
```

Comment:
It's difficult to understand what is going on in this test.  After reading it a couple times, it looks like you're setting up a fake server which will get three requests.  Every time that server gets one of the requests, it will first verify that the UI is in the expected state, send the response based upon the array, again verify that the UI is in the expected state, and then update the UI to have it issue another request to itself.

That took a lot of reading and I think it would be useful to define a couple helper methods to deal with components of that.  If the methods have descriptive enough names (e.g. `_verifyWidgetsVisible(mode)`) it may be possible to do this without a large number of comments.

I'm also a bit confused at how demoSettings is used in this case...from the variable name, I'm assuming this is something that would not change while the test is ongoing, but there are a lot of places in the test that seem dependent upon the value of the settings.  Can you explain that a bit more?

I don't have enough familiarity with the actual code here or the tests to say exactly how to make it more clear, but I get the feeling that it's possible.  Happy to sit down if you have time to explain more! (this comment also applies for the next couple tests)

Reply:
Yes, you have it correct. I could have made this clearer--I've been lax on documentation since it makes sense to me, but I realize now that it's not clear at all to anyone else. Thank you for the feedback, I will try to improve the documentation. I totally understand what you're trying to say--these tests are full of clunky logic and I'll do what I can to simplify them into sub-methods.

Regarding the demo settings not changing--you are right. What happened was I originally had this all in one giant messy test, but then I moved it into separate tests and copy/pasted some of the logic, but looks like I missed deleting some of the case logic that was no longer needed. I will go and clean that up.